### DISCUSSION TITLE
How to request a new access token with additional extraTokenParameters after the user has already authenticated?

### DISCUSSION THREAD

**Discussion by jessgoldq4** [soofstad/react-oauth2-pkce#146] [Q&A]
Hi there. Question. I want to retrieve a new access token for a user with a new set of custom claims after they have already logged in and retrieved their initial access token. 

Is there something built into the package to enable us to force a new access token to be generated? Or do I need to hit our IDP's refresh token endpoints directly?

What's the best way to accomplish this with react-oauth2-pkce? 

Thanks in advance

**Comment by soofstad** [comment]
Hi,
There is no feature to call the token endpoint at will.
Also, if you change the claim, the IDP should deny the request for a new access token using the old refresh token.
That means you would need to login from scratch, with the new claim.

This flow you are describing definitely seems a bit out of the ordinary. I could try to help out if you provide some more information :)

**Reply by jessgoldq4** [reply]
Interesting. This functionality is actually possible with Auth0's react sdk out of the box by using their provided `getAccessTokenSilently()` [function](https://github.com/auth0/auth0-react/blob/6bf58e8b89f3ed49b1fbf701aee7d9088f4ccedb/src/auth0-provider.tsx#L209-L226) as noted in their [comments](https://github.com/auth0/auth0-react/blob/6bf58e8b89f3ed49b1fbf701aee7d9088f4ccedb/src/auth0-context.tsx#L26-L62):

>    \* If there's a valid token stored, return it. Otherwise, opens an
      * iframe with the `/authorize` URL using the parameters provided
      * as arguments. Random and secure `state` and `nonce` parameters
     * will be auto-generated. If the response is successful, results
     * will be valid according to their expiration times.
     *
     * If refresh tokens are used, the token endpoint is called directly with the
     * 'refresh_token' grant. If no refresh token is available to make this call,
     * the SDK will only fall back to using an iframe to the '/authorize' URL if 
     * the `useRefreshTokensFallback` setting has been set to `true`. By default this
     * setting is `false`.  
       ----> ((jessgoldq4 note: we have useRefreshTokensFallback set to false))
     *
     * This method may use a web worker to perform the token call if the in-memory
     * cache is used. 
        ---> ((jessgoldq4 note: we are ignoring the in-memory cache))
     *
     * If an `audience` value is given to this function, the SDK always falls
     * back to using an iframe to make the token exchange.
     *
     * Note that in all cases, falling back to an iframe requires access to
     * the `auth0` cookie.
     */

Lower level `getTokenSilently()` implementation [here](https://github.com/auth0/auth0-spa-js/blob/8653181d23ecea17729da203253d4fa4d17d8ef6/src/Auth0Client.ts#L643-L664)

With use of this function, I'm can pass in some options to set some values, which is equivalent to setting some values in the react-oauth2-pkce `extraTokenParameters`. And by doing so, I'm able to retrieve a new set of id and access tokens that contain new custom claims based on the values I've set in the getAccessTokenSilently() options argument. This does not require the user to re-enter their username and password again. If that's what you meant by needing to login from scratch, I don't believe that's necessary

i.e,

```
  useEffect(() => {
    const options = { ignoreCache: true, sampleExtraTokenParameterValue: sampleExtraTokenParameterValue || undefined };

    auth0
      .getAccessTokenSilently(options)
      .then(() => auth0.getIdTokenClaims())
      .then((newClaims) => {
        setClaims(newClaims);
        setIsSampleExtraTokenParameterValueLoading(false);
      });
  }, [auth0, sampleExtraTokenParameterValue]);
  ```

Any recommendations on how to accomplish this by using `react-oauth2-pkce`? Your help would be greatly appreciated!

**Reply by soofstad** [reply]
Right, I see how they solved it. So technically, we can add the same functionality.
But from the [RFC6749](https://datatracker.ietf.org/doc/html/rfc6749#section-6)
> scope
         OPTIONAL.  The scope of the access request as described by
         [Section 3.3](https://datatracker.ietf.org/doc/html/rfc6749#section-3.3).  The requested scope MUST NOT include any scope
         not originally granted by the resource owner, and if omitted is
         treated as equal to the scope originally granted by the
         resource owner.

This means you can not use the refresh_token to do a "silent_background_fetch" if you change the claim.

It is possible to do a new silent login (as they do) to the /auth endpoint, but that only works if the IDP uses a cookie to persist authentication across sessions, if not, the user would need to reenter username and password.

If the library is to support this, the first step would be to add support for iframe or popup login. That might not be a bad idea.

A similar (but not quite) feature, would be #145. When this is added, you can pass your updated claim to the login function and call that. It will do a redirect (and send the user right back with no need to enter credentials if the IDP has set a cookie), but it will not be "silent in the background".

I'm working on this issue now, so when it's ready you can test it and see if it achieves what you want. If not, an identical feature to that of Auth0 will take longer to implement.

**Reply by jessgoldq4** [reply]
> This means you can not use the refresh_token to do a "silent_background_fetch" if you change the claim.

I'm not sure what you mean when you say 'if you change the claim'. 

To clarify, in the example I described, I want to pass in an extra parameter in the payload of the `oauth/token` request. In Auth0, you can configure scripts that run after certain actions. In this case, I have an Auth0 `post-login` action written that will insert custom claims into the id and access tokens before they're returned to the user if the extra parameter is received. 

With that said, what exactly do you mean by changing the claim here? Not sure I'm following exactly

> It is possible to do a new silent login (as they do) to the /auth endpoint, but that only works if the IDP uses a cookie to persist authentication across sessions, if not, the user would need to reenter username and password.

Hmm, interesting. I would assume that's the case for Auth0 then, because my users are not asked to re-enter their credentials after calling `getAccessTokenSilently({ ignoreCache: true, sampleExtraTokenParameterValue: sampleExtraTokenParameterValue || undefined })`

>A similar (but not quite) feature, would be https://github.com/soofstad/react-oauth2-pkce/issues/145. When this is added, you can pass your updated claim to the login function and call that.

Thanks for sharing this! Initial tests seems to show this doesn't work to solve the issue, but I'll take another look shortly

>It will do a redirect (and send the user right back with no need to enter credentials if the IDP has set a cookie), but it will not be "silent in the background".

That aligns with my expectation. When a new token is issued via `getAccessTokenSilently()`, the application will refresh, and it won't be silent. This is how my application currently works

>If not, an identical feature to that of Auth0 will take longer to implement.

Hmm. Maybe I can try my hand at implementing this on a fork of `react-oauth2-pkce` as a POC. What do you think?

**Reply by jessgoldq4** [reply]
Actually, it seems like the functionality I need is working as intended by calling login() a second time with an extra param passed in! Amazing! 

Might not need anything else done here to accommodate this need. Will confirm and update shortly :)

Thanks so much

**Reply by soofstad** [reply]
That's good to hear. Let me know how it goes :slightly_smiling_face:

**Reply by abdullahmonoar** [reply]
Hi @soofstad, thanks for the quick turn around on the update. I think it mostly does what we want, but not entirely in the experience that we're looking for.

So we have a feature in our app, where you click on a user to impersonate, and we want to refresh the token in the background. If we call login, this does a full page refresh and not really what we want. Is there a way to expose a method so that we can do this refresh in the background with additional params?

We've got a bit of a deadline for this change, end of May. We're happy to support in any way we can, do you think it would be possible to prioritize this functionality?

for reference, we're trying to replicate this behaviour: https://github.com/auth0/auth0-spa-js/blob/8653181d23ecea17729da203253d4fa4d17d8ef6/src/Auth0Client.ts#L643

**Reply by soofstad** [reply]
I see. It would be a useful feature. Most of the infrastructure is already there. What is lacking is the `makingAuthRequestInIFrame`-function. That would create cause cross-origin issues for most setups, that you would then need to fix.

To develop and test this properly would take some time. I can not promise I will get around to it anytime soon. I've created an issue for it (#147). If any of you are able to contribute with this that would be much appreciated.

**Reply by abdullahmonoar** [reply]
Okay very nice, I think you're definitely more versed with Authentication. Wondering if one solution could be creating a function like so in AuthContext and what security concerns this poses?

```
  function loginSilently(tokenParams: TPrimitiveRecord = {}) {
    fetchTokens({...config, extraTokenParameters: tokenParams })
      .then((tokens: TTokenResponse) => {
        clearStorage()
        handleTokenResponse(tokens)
      })
      .catch((error: Error) => {
        console.error(error)
        setError(error.message)
      })
  }
  ```

**Reply by soofstad** [reply]
That function only fetches tokens using a `code` or a refresh token. That can happen with a post request.
What you need is a redirect, similar to https://github.com/soofstad/react-oauth2-pkce/blob/main/src/authentication.ts#L15, but in an iframe or popup
